contrib/pkg/awsprivatelink/endpointvpc/add.go

```diff
@@ -0,0 +1,469 @@
+					Name:   aws.String("association.subnet-id"),
+					Values: aws.StringSlice(endpointSubnetIds),
+				},
+			},
+		},
+		func(page *ec2.DescribeRouteTablesOutput, lastPage bool) bool {
+			for _, routeTable := range page.RouteTables {
+				_, err := endpointVpcEC2Client.CreateRoute(&ec2.CreateRouteInput{
+					RouteTableId:           routeTable.RouteTableId,
+					DestinationCidrBlock:   associatedVpcCIDR,
+					VpcPeeringConnectionId: VpcPeeringConnectionId,
+				})
+				if err != nil {
+					// Proceed if route already exists, fail otherwise
+					switch aerr, ok := err.(awserr.Error); {
+					case ok && aerr.Code() == "RouteAlreadyExists":
+						log.Warnf("Route already exists in route table %v", *routeTable.RouteTableId)
+					default:
+						log.WithError(err).Fatalf("Failed to create route for route table %v", *routeTable.RouteTableId)
+					}
+				} else {
+					log.Debugf("Route added to route table %v", *routeTable.RouteTableId)
+				}
+			}
+
+			return !lastPage
+		},
+	)
+}
+
+func addRouteToPrivateRouteTablesOfAssociatedVpc(
+	associatedVpcClients awsclient.Client,
+	associatedVpcId, endpointVpcCIDR, VpcPeeringConnectionId *string,
+) error {
+	log.Info("Adding routes to private route tables of the associated VPC")
+
+	return associatedVpcClients.DescribeRouteTablesPages(
+		&ec2.DescribeRouteTablesInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("vpc-id"),
+					Values: []*string{associatedVpcId},
+				},
+				{
+					Name:   aws.String("tag:Name"),
+					Values: []*string{aws.String("*private*")},
+				},
+			},
+		},
+		func(page *ec2.DescribeRouteTablesOutput, lastPage bool) bool {
+			for _, routeTable := range page.RouteTables {
+				_, err := associatedVpcClients.CreateRoute(&ec2.CreateRouteInput{
+					RouteTableId:           routeTable.RouteTableId,
+					DestinationCidrBlock:   endpointVpcCIDR,
+					VpcPeeringConnectionId: VpcPeeringConnectionId,
+				})
+				if err != nil {
+					// Proceed if route already exists, fail otherwise
+					switch aerr, ok := err.(awserr.Error); {
+					case ok && aerr.Code() == "RouteAlreadyExists":
+						log.Warnf("Route already exists in route table %v", *routeTable.RouteTableId)
+					default:
+						log.WithError(err).Fatalf("Failed to create route for route table %v", *routeTable.RouteTableId)
+					}
+				} else {
+					log.Debugf("Route added to route table %v", *routeTable.RouteTableId)
+				}
+			}
+
+			return !lastPage
+		},
+	)
+}
+
+// Does not error out even if the peering connection is already established between the two VPCs.
+func setupVpcPeeringConnection(
+	associatedVpcClients, endpointVpcClients awsclient.Client,
+	associatedVpcId, endpointVpcId, endpointVpcRegion *string,
+) (*ec2.AcceptVpcPeeringConnectionOutput, error) {
+	log.Info("Setting up VPC peering connection between the associated VPC and the endpoint VPC")
+
+	createVpcPeeringConnectionOutput, err := associatedVpcClients.CreateVpcPeeringConnection(&ec2.CreateVpcPeeringConnectionInput{
```

Comment:
Is this idempotent? I.e. succeeds if the connection already exists?

[Later] Looks like it's an async call whose error is only non-nil if, like, the network bit fails. So :+1: